### PR TITLE
Immunization Resource with GET method

### PR DESCRIPTION
**Pull request by silvistershebin**
Fixes Part 1 of #3346

#### Short description:
Immunization Resource with Get Methods:
    - getAll
    - getOne
    - get Immunization by patient id

### CODE REVIEW

**Comment by sjpadgett** [comment]
I do believe you've got the hang of this whole resource thing @silvistershebin :) After three years of begging for help with FHIR and to see a new resource, makes me giddy:)
Anyway, looks great but, i'll review closer after my third cup of coffee!

File: src/Services/FHIR/FhirImmunizationService.php [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        return $result;        
+    }
+
+    public function createImmunizationResource($id = '', $data = '', $encode = true)
+    {     
+        $status = new FHIRImmunizationStatusCodes();
+        $statusCoding = new FHIRCoding();
+        if (isset($data['completion_status'])) {
+            $statusCoding = new FHIRCoding();
+            $statusCoding->setSystem('http://hl7.org/fhir/event-status');
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:106]: I think this url can change to https

File: src/Services/FHIR/FhirImmunizationService.php [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            $statusCoding->setCode($data['completion_status']);
+            $statusCoding->setDisplay($data['completion_status']);
+            $status->setValue($statusCoding);
+        }
+        else {
+            $statusCoding->setSystem("http://terminology.hl7.org/CodeSystem/data-absent-reason");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:112]: This url seems broken.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            $statusCoding->setCode("unknown");
+            $statusCoding->setDisplay("Unknown");        
+            $status->setValue($statusCoding);
+        }
+        
+        $statusReason = new FHIRCodeableConcept();
+        $statusReasonCoding = new FHIRCoding();
+        $statusReasonCoding->setSystem("http://terminology.hl7.org/CodeSystem/v3-ActReason");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:120]: This url seems broken.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                                    "product out of stock" => "OSTOCK",
+                                    "patient objection" => "PATOBJ"
+                                );
+        $statusReasonCoding->setCode($statusToDisplayMapping[strtolower($data['refusal_reason'])]);
+        $statusReasonCoding->setDisplay($data['refusal_reason']);
+        $statusReason->addCoding($statusReasonCoding);
+        
+        $vaccineCode = new FHIRCodeableConcept();
+        $vaccineCodeCoding = new FHIRCoding();
+        $vaccineCodeCoding->setSystem("http://hl7.org/fhir/sid/cvx");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:133]: This url seems broken.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        
+        $lotNumber = new FHIRString();
+        $lotNumber->setValue($data['lot_number']);
+        
+        $expirationDate = new FHIRDate();
+        $expirationDate->setValue($data['expiration_date']);
+        
+        $site = new FHIRCodeableConcept();
+        $siteCoding = new FHIRCoding();
+        $siteCoding->setSystem("http://terminology.hl7.org/CodeSystem/v3-ActSite");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:158]: This url seems broken.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        $siteCoding->setCode($data['administration_site']);
+        $siteCoding->setDisplay($data['administration_site']);
+        $site->addCoding($siteCoding);
+        
+        $route = new FHIRCodeableConcept();
+        $routeCoding = new FHIRCoding();
+        $routeCoding->setSystem("http://terminology.hl7.org/CodeSystem/v3-RouteOfAdministration");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:165]: This url seems broken.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        $routeCoding->setCode($data['route']);
+        $routeCoding->setDisplay($data['route']);
+        $route->addCoding($routeCoding);
+
+        $doseQuantity = new FHIRQuantity();
+        $doseQuantity->setValue($data['amount_administered']);
+        $doseQuantity->setSystem($data['http://unitsofmeasure.org']);
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:172]: I think this url can change to https

File: src/Services/FHIR/FhirImmunizationService.php [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        $doseQuantity->setCode($data['amount_administered_unit']);
+        
+        $actor = new FHIRReference();
+        $actor->setReference($data['administered_by'].'/'.$data['administered_by_id']);
+        $function = new FHIRCodeableConcept();
+        $functionCoding = new FHIRCoding();
+        $functionCoding->setSystem("http://terminology.hl7.org/CodeSystem/v3-RouteOfAdministration");
```
Comment (stu01509) [inline src/Services/FHIR/FhirImmunizationService.php:179]: This url seems broken.

**Comment by sjpadgett** [comment]
@rhausam 
I don't think FHIR maintains a public terminology server. Everywhere I look I see these references being used.
Should we create a table of valuesets and only add sets that we are using? Do we even have to supply a ref to the value set but instead only the valueset type?

**Comment by yashrajbothra** [comment]
> @rhausam
> I don't think FHIR maintains a public terminology server. Everywhere I look I see these references being used.
> Should we create a table of valuesets and only add sets that we are using? Do we even have to supply a ref to the value set but instead only the valueset type?

I have few ideas for valuesets will try once this Observation Resource is completed :)

**Comment by rhausam** [comment]
FHIR does maintain tx.fhir.org for the FHIR and Implementation Guide (IG) build processes.  And HL7 is also standing up terminology.hl7.org (also known as "UTG"), ultimately as a repository for all HL7 code systems and value sets.  And each FHIR IG can host its own additional terminology content as needed.

Most likely a strategy of using internal tables to cache the value set contents (specifically the value set **expansion** contents) for the particular value sets that are needed for OpemEMR functionality, coupled with a reasonable and reliable strategy for updating the internal tables as the content in the "official" sources changes, would be the best approach.  Maybe first we need to see if we have or can compile an inventory of the value sets that are needed, and see how that looks.  I'm not exactly sure what you mean by "Do we even have to supply a ref to the value set but instead only the valueset type" - maybe you can clarify that?

And probably we should move this value set discussion somewhere besides "Immunization Resource with GET method"?

**Comment by sjpadgett** [comment]
on you @yashrajbothra as this is part of your project. Start an issue or you may have one already.
I'll discuss further there.

File: src/Services/FHIR/FhirImmunizationService.php [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+    public function createImmunizationResource($id = '', $data = '', $encode = true)
+    {     
+        $status = new FHIRImmunizationStatusCodes();
+        $statusCoding = new FHIRCoding();
+        if (isset($data['completion_status'])) {
+            $statusCoding = new FHIRCoding();
+            $statusCoding->setSystem('http://hl7.org/fhir/event-status');
+            $statusCoding->setCode($data['completion_status']);
+            $statusCoding->setDisplay($data['completion_status']);
```
Comment (sjpadgett) [inline src/Services/FHIR/FhirImmunizationService.php:108]: For now I think we need to translate our values to match hl7 value sets.
i.e in-progress | on-hold | completed | entered-in-error | stopped

File: src/Services/FHIR/FhirImmunizationService.php (hunk 4 above)
Comment (sjpadgett) [reply src/Services/FHIR/FhirImmunizationService.php:133]: I'm not worried if valid url to terminology build for now.

**Comment by silvistershebin** [comment]
Hi @sjpadgett Sir, I have updated the code. Kindly review it.

**Pull request merged**